src/v/kafka/server/handlers/fetch.cc

```diff
@@ -1176,6 +1180,21 @@ class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
         }
 
         const bool foreign_read = fetch.shard != ss::this_shard_id();
+        const auto bytes_left = octx.bytes_left;
+
+        ss::abort_source as{};
+        auto resp_order = ss::now();
+        if (is_initial_request) {
+            // Ensure that initial shard fetch responses are filled in the same
+            // order that the shard workers are started in. This is to prevent
+            // the non-foreign shard worker from being preferred as it is more
+            // likely than the rest to be the first worker the executor gets
+            // results for.
+            resp_order = _initial_response_ordering.wait();
+            // Prevent the workers from waiting for new partition data on the
+            // initial request.
+            as.request_abort();
```

Comment:
I don't understand this whole block. How I would imagine the solution to work is that somehow we sort the response based on the order in the requests. Aren't you just fully serializing the different shards here like this? That doesn't seem intentional to me. 

If it is I'd rather just expect a serial outer loop rather than this semaphore/abort_source pattern.